function-controller/pkg/controller/controller.go

```diff
@@ -354,111 +351,39 @@ func New(topicInformer informersV1.TopicInformer,
 		}()
 	}
 
-	return pctrl
-}
-
-func DecorateWithDelayAndSmoothing(c Controller) {
-	pctrl := c.(*ctrl)
-	smoother := smoother{ctrl: pctrl, memory: make(map[fnKey]float32)}
-	pctrl.scalingPolicy = compose(smoother.smooth, pctrl.scalingPolicy)
-
-	delayer := delayer{ctrl: pctrl, scaleDownToZeroDecision: make(map[fnKey]time.Time), previousCombinedPositions: make(activityCounts)}
-	pctrl.scalingPolicy = compose(delayer.delay, pctrl.scalingPolicy)
-}
-
-// computeDesiredReplicas turns a subscription based map (of offsets) into a function-key based map of how many replicas to spawn.
-// The logic is as follows: for a given topic, look at how many partitions have lag or activity (there is no point in spawning 2
-// replicas if only a single partition has lag, even if it's a 1000 messages lag).
-// If the function has multiple inputs, we take the max of those computations (some topics may be starved but that's ok
-// while we still maximize the throughput for the given topic that has the most needy partitions)
-func (c *ctrl) computeDesiredReplicas(offsets map[Subscription]PartitionedOffsets) (replicaCounts, activityCounts) {
-	replicas := make(replicaCounts)
-	combinedCurrentPositions := make(activityCounts)
-	for s, o := range offsets {
-		fn := fnKey{s.Group}
-		topic := topicKey{s.Topic}
-		replicas[fn] = max(replicas[fn], c.desiredReplicasForTopic(o, fn, topic))
-		ccp := combinedCurrentPositions[fn]
-		ccp.end += sumEndPositions(o)
-		ccp.current += sumCurrentPositions(o)
-		combinedCurrentPositions[fn] = ccp
-	}
-	return replicas, combinedCurrentPositions
-}
-
-func sumEndPositions(offsets PartitionedOffsets) int64 {
-	result := int64(0)
-	for _, o := range offsets {
-		result += o.End
-	}
-	return result
-}
-
-func sumCurrentPositions(offsets PartitionedOffsets) int64 {
-	result := int64(0)
-	for _, o := range offsets {
-		result += o.Current
-	}
-	return result
-}
-
-func (c *ctrl) desiredReplicasForTopic(offsets PartitionedOffsets, fnKey fnKey, tKey topicKey) int32 {
-	maxPartLag := int64(0)
-	for _, o := range offsets {
-		maxPartLag = max64(maxPartLag, o.Lag())
-	}
-
-	// TODO: those 3 numbers part of Function spec?
-	lagRequiredForMax := float32(10)
-	lagRequiredForOne := float32(1)
-	minReplicas := int32(0)
+	auto.SetMaxReplicasPolicy(func(topic string, function string) int {
```

Comment:
I don't really see the value of this being pluggable

Reply:
I wanted to avoid the autoscaler needing to deal with topic and function configuration. I imagine the policy being different for non-Kafka transports which don't have partitions.

Reply:
Ah, good point (although our very topic CRD has the concept of partitions, for now)